content/topics/kotlin/generics/_index.md

```diff
@@ -6,16 +6,342 @@ prerequisites:
+
+### Declaration-site variance
+Suppose we have a generic interface Source<T> that does not have any methods that take T as a parameter, only methods that return T:
+```
+// Java
+interface Source<T> {
+  T nextT();
+}
+```
+Then, it would be perfectly safe to store a reference to an instance of Source<String> in a variable of type Source<Object> – there are no consumer-methods to call. But Java does not know this, and still prohibits it:
+```
+// Java
+void demo(Source<String> strs) {
+  Source<Object> objects = strs; // !!! Not allowed in Java
+  // ...
+}
+```
+To fix this, we have to declare objects of type 
+
+```
+Source<? extends Object>
+```
+which is sort of meaningless, because we can call all the same methods on such a variable as before, so there's no value added by the more complex type. But the compiler does not know that.
+
+In Kotlin, there is a way to explain this sort of thing to the compiler. This is called declaration-site variance: we can annotate the type parameter T of Source to make sure that it is only returned (produced) from members of Source<T>, and never consumed. To do this we provide the out modifier:
+```
+interface Source<out T> {
+    fun nextT(): T
+}
+​
+fun demo(strs: Source<String>) {
+    val objects: Source<Any> = strs // This is OK, since T is an out-parameter
+    // ...
+}
+```
+The general rule is: when a type parameter T of a class C is declared out, it may occur only in out-position in the members of C, but in return C<Base> can safely be a supertype of C<Derived>.
+
+In "clever words" they say that the class C is covariant in the parameter T, or that T is a covariant type parameter. You can think of C as being a producer of T's, and NOT a consumer of T's.
+
+The out modifier is called a variance annotation, and since it is provided at the type parameter declaration site, we talk about declaration-site variance. This is in contrast with Java's use-site variance where wildcards in the type usages make the types covariant.
+
+In addition to out, Kotlin provides a complementary variance annotation: in. It makes a type parameter contravariant: it can only be consumed and never produced. A good example of a contravariant type is Comparable:
+```
+interface Comparable<in T> {
+    operator fun compareTo(other: T): Int
+}
+​
+fun demo(x: Comparable<Number>) {
+    x.compareTo(1.0) // 1.0 has type Double, which is a subtype of Number
+    // Thus, we can assign x to a variable of type Comparable<Double>
+    val y: Comparable<Double> = x // OK!
+}
+```
+We believe that the words in and out are self-explaining (as they were successfully used in C# for quite some time already), thus the mnemonic mentioned above is not really needed, and one can rephrase it for a higher purpose:
+
+The Existential Transformation: Consumer in, Producer out! :-)
+
+### Type projections
+####Use-site variance: Type projections
```

Comment:
There should be a space here as it doesnt render as a heading

Reply:
Done

Reply:
looks like it may have been missed?